registry/client/auth/session.go

```diff
@@ -32,71 +34,24 @@ type CredentialStore interface {
 
 // NewAuthorizer creates an authorizer which can handle multiple authentication
 // schemes. The handlers are tried in order, the higher priority authentication
-// methods should be first.
-func NewAuthorizer(transport http.RoundTripper, handlers ...AuthenticationHandler) RequestModifier {
-	return &tokenAuthorizer{
-		challenges: map[string]map[string]authorizationChallenge{},
+// methods should be first. The challengeMap holds a list of challenges for
+// a given root API endpoint (for example "https://registry-1.docker.io/v2/").
+func NewAuthorizer(challengeMap map[string][]Challenge, handlers ...AuthenticationHandler) transport.RequestModifier {
```

Comment:
That might need some work. Why not have it take a challenge response?

```
func NewAuthorizer(challengeResponse *http.Response, handlers ...AuthenticationHandler) transport.RequestModifier
```

This still feels awkward...

Maybe hide it in the library?

```
_, err := auth.Check(resp, "")

if challenge, ok := err.(Challenge); ok {
    authorizer := challenge.Authorizor()
}
```